### PR TITLE
fix(mdPanel): Propagation, CSS targeting, and dynamic position updating

### LINKED ISSUE #8968: mdPanel: Implement "propagateContainerClicks" config option.
This option should allow users to interact with other elements on the page when the panel is opened.

If `clickOutsideToClose` is also present, the panel should close itself and then click on the underlying element.

If `hasBackdrop` is true, this option should be ignored as the backdrop will visually indicate that clicks will stop at that barrier.

This is currently not possible because the panel container covers the entire screen and captures any click events.

This feature is needed for tooltips and menus.

### LINKED ISSUE #8980: fix(mdPanel): Dynamic update of panel position and CSS class manipulation
**Actual Behavior**:
 -  `What is the issue? *` **This is actually a two-part issue**
  - **Part 1:** There is no current way to update the position of the mdPanel that is originally established in the configuration, without closing the mdPanel, changing the position and re-opening the mdPanel.
  - **Part 2:** The `addClass`, `removeClass`, and `toggleClass` functions of the `MdPanelRef` only target the mdPanel container, the 'outer-wrapper', even though a developer might want to target the mdPanel element instead.

 -  `What is the expected behavior?`
  - **Part 1:** There should be a public function of the mdPanel API that allows you to update the position object in the current panel's configuration and then it should update the position of the mdPanel.
  - **Part 2:** There should be an option on the `addClass`, `removeClass`, and `toggleClass` functions that allow you target the mdPanel element instead of the container.
  - **Part 2 (EDIT):** **An example of why a developer would want to use** the `addClass`, `removeClass`, and `toggleClass` modifications to specifically target the panel element:
    - Targeting the panel element is useful if you want to add CSS classes to the panel element for custom transitions or behavior. For instance, the transition between showing and hiding the panel with an animation that is not supported by the mdPanelAnimation API.

**CodePen** (or steps to reproduce the issue): *
 -  `CodePen Demo which shows your issue:` N/A
 -  `Details:` N/A

**Angular Versions**: *
 -  `Angular Version:` Latest
 -  `Angular Material Version:` Latest

**Additional Information**:
 -  `Browser Type: *` All
 -  `Browser Version: *` All
 -  `OS: *` All
 -  `Stack Traces:` N/A

### PR DESCRIPTION
**Pull request by bradrich**
Fixes #8968 by:
 - Adding the `propagateContainerEvents` option to the panel configuration that is defaulted to false. If true, the mdPanel API will allow all touch and pointer events to propagate through the mdPanel container, the 'outer-wrapper', to the elements behind it.

Fixes #8980 by:
- Adding a boolean parameter to the MdPanelRef `addClass`, `removeClass`, and `toggleClass` functions that defaults to false. When true, it will allow the function's primary class actions to be executed on the mdPanel element, instead of the container.
- Adding a public `updatePosition` function that will take in a MdPanelPosition object parameter that it will overwrite the current configuration's position and then call the private `_updatePosition` function.

Ping @ErinCoughlan @EladBezalel

### CODE REVIEW

File: src/components/panel/panel.js [hunk 1]
```diff
@@ -1188,6 +1211,24 @@ MdPanelRef.prototype._addStyles = function() {
 
 
 /**
+ * Update the position configuration of a panel
```
Comment (ErinCoughlan) [inline src/components/panel/panel.js]: nit: Method descriptions should start with a sentence written in the third person declarative voice. Can you change this to:

"Updates the position configuration of the panel."

File: src/components/panel/panel.js [hunk 2]
```diff
@@ -1188,6 +1211,24 @@ MdPanelRef.prototype._addStyles = function() {
...
+ * @param {MdPanelPosition} position
+ */
+MdPanelRef.prototype.updatePosition = function(position) {
```
Comment (ErinCoughlan) [inline src/components/panel/panel.js]: You need to add this documentation also to the public documentation above, probably with a description of why you someone would use this. (~line 330)

File: src/components/panel/panel.js [hunk 3]
```diff
@@ -1086,13 +1098,19 @@ MdPanelRef.prototype.removeClass = function(oldClass) {
  * Toggle a class on the panel. DO NOT use this to hide/show the panel.
  *
  * @param {string} toggleClass The class to toggle.
+ * @param {boolean} onElement Whether or not to toggle the class on the panel
```
Comment (ErinCoughlan) [inline src/components/panel/panel.js]: New params need to be added to the public documentation at the top of this file. (For add, remove, and toggle)

**Comment by bradrich** [comment]
@ErinCoughlan Thank you for these comments. I will make your suggested changes.

**Comment by EladBezalel** [comment]
IMO we should remove addClass, removeClass and toggleClass by providing the container and the panel elements as a public api. 
Doesn't makes sense copying this functionality from jqlite. 
@ErinCoughlan

**Comment by ErinCoughlan** [comment]
@EladBezalel That's fine by me.

**Comment by bradrich** [comment]
@topherfangio Squashed my commits on this, even though it needs work.

**Comment by topherfangio** [comment]
@bradrich Awesome; thanks! In general, if you push anything to a PR (even in-progress), you should squash your commits 😄

**Comment by ErinCoughlan** [comment]
@bradrich Tests? You added new functionality, but I have no idea if they work and will continue to work.

**Comment by bradrich** [comment]
@ErinCoughlan All tests are included and passing. Please review and give LGTM.

File: src/components/panel/panel.spec.js [hunk 4]
```diff
@@ -1115,6 +1212,63 @@ describe('$mdPanel', function() {
       mdPanelPosition = $mdPanel.newPanelPosition();
     });
 
+    it('should update the position of an open panel', function() {
```
Comment (bradrich) [inline src/components/panel/panel.spec.js]: @ErinCoughlan Please give special attention to this test.

File: src/components/panel/panel.spec.js [hunk 5]
```diff
@@ -868,6 +895,76 @@ describe('$mdPanel', function() {
       });
     });
 
+    describe('CSS class logic:', function() {
+      it('should add a class to the container/wrapper when toElement=false',
+        function() {
+        openPanel(DEFAULT_CONFIG);
```
Comment (ErinCoughlan) [inline src/components/panel/panel.spec.js]: nit: indentation here is off. openPanel and the rest of the code still needs to be indented under function, like this:
```
it('should add a class to the container/wrapper when toElement=false',
  function() {
    openPanel(DEFAULT_CONFIG);
    ...
    expect(PANEL_WRAPPER_CLASS).toHaveCLASS('my-class');
  });

File: src/components/panel/panel.spec.js [hunk 6]
```diff
@@ -868,6 +895,76 @@ describe('$mdPanel', function() {
...
+
+        panelRef.addClass('my-class');
+
+        expect(PANEL_WRAPPER_CLASS).toHaveClass('my-class');
```
Comment (ErinCoughlan) [inline src/components/panel/panel.spec.js]: Let's add another expect here to be safe. Technically, the class can be add to the wrapper AND element and still pass all of these tests. Same with the below tests.

```
expect(PANEL_WRAPPER_CLASS).toHaveClass('my-class');'
expect(PANEL_EL).not.toHaveClass('my-class');
```

File: src/components/panel/panel.spec.js [hunk 7]
```diff
@@ -1115,6 +1212,62 @@ describe('$mdPanel', function() {
       mdPanelPosition = $mdPanel.newPanelPosition();
     });
 
+    it('should update the position of an open panel', function() {
+      var position = null;
+
+      // Starting off with a absolute positioned panel at the top of the
+      // window.
+      testAbsTop();
```
Comment (ErinCoughlan) [inline src/components/panel/panel.spec.js]: Each one of these should be a different test. It's hard to follow what the before and after states of each test is with this format. In tests, you want to avoid someone reviewing it having to think. I should be able to look at the test and say "Okay, there's the setup. Some action was taken. And the values changed correctly. Done."

Rarely should you need comments to explain what is happening. That's generally means the test can be better organized (functions that run asserts are generally bad).

Possibly format like this:

```
describe('should update the position of an open panel', function() {
  it('between two relative positions', function() { 
    // Create a relatively positioned panel.
    // Verify the position is initially right.
    // Call updatePosition with new relative position object.
    // Verify the position was updated.
  });

  it('between two absolute positions', function() {
    // Create an absolutely positioned panel.
    // Verify the position is initially right.
    // Call updatePosition with new absolute position object.
    // Verify the position was updated.
  });

  it('from absolute positioning to relative positioning', function() { ... });
});
```

**Comment by bradrich** [comment]
@ErinCoughlan That's it! Whew!

File: src/components/panel/panel.js [hunk 8]
```diff
@@ -1194,6 +1232,24 @@ MdPanelRef.prototype._addStyles = function() {
...
+ */
+MdPanelRef.prototype.updatePosition = function(position) {
+  if (!this._panelContainer) {
+    throw new Error('Panel does not exist yet. Call open() or attach().');
+  }
+
+  var self = this;
+
+  self._config['position'] = position;
+  self._$mdUtil.nextTick(function() {
```
Comment (ErinCoughlan) [inline src/components/panel/panel.js]: Does this work?

```self._$mdUtil.nextTick(self._updatePosition);```

Comment (DevVersion) [inline src/components/panel/panel.js]: I guess it would work if you bind it to `this`, since `updatePosition` accessed the instance context?

Comment (bradrich) [inline src/components/panel/panel.js]: It does not. I have new tests coming in.

**Comment by bradrich** [comment]
@EladBezalel I am going to add the panel wrapper/container and element to the panelRef in another feature. So, I am going to remove the needs: work label.

**Comment by bradrich** [comment]
@ThomasBurleson There is an issue with the Travis CI. Is it known?

**Comment by DevVersion** [comment]
@bradrich This rarely happens to builds on Saucelabs. I will try to increase the Sauce timeout, so we reduce that flakiness a bit.

File: src/components/panel/panel.js [hunk 9]
```diff
@@ -1194,6 +1232,22 @@ MdPanelRef.prototype._addStyles = function() {
...
 /**
+ * Updates the position configuration of a panel
+ * @param {MdPanelPosition} position
+ */
+MdPanelRef.prototype.updatePosition = function(position) {
+  if (!this._panelContainer) {
+    throw new Error('Panel does not exist yet. Call open() or attach().');
+  }
+
+  var self = this;
```
Comment (ErinCoughlan) [inline src/components/panel/panel.js]: I think the self variable here is no longer necessary and you can use `this` for both lines below.

File: src/components/panel/panel.spec.js [hunk 10]
```diff
@@ -409,6 +409,34 @@ describe('$mdPanel', function() {
       });
     });
 
+    describe('should cause the propagation of events', function() {
+      var config, wrapper;
+
+      it('to be stopped when propagateContainerEvents=false', function() {
+        config = {
+          propagateContainerEvents: false,
+          template: '<div style="height: 300px">Hello World</div>'
```
Comment (ErinCoughlan) [inline src/components/panel/panel.spec.js]: Is the template still necessary here? Or does `DEFAULT_TEMPLATE` work?

File: src/components/panel/panel.spec.js [hunk 11]
```diff
@@ -868,6 +895,87 @@ describe('$mdPanel', function() {
...
+        expect(PANEL_WRAPPER_CLASS).not.toHaveClass('my-class');
+      });
+
+      it('should remove a class from the container/wrapper when fromElement=false',
+        function() {
+          openPanel(DEFAULT_CONFIG);
+
+          panelRef.addClass('my-class');
+
+          expect(PANEL_WRAPPER_CLASS).toHaveClass('my-class');
```
Comment (ErinCoughlan) [inline src/components/panel/panel.spec.js]: super nit: Can you keep the order of these the same for all the tests in this describe? Either put `PANEL_WRAPPER_EL` expect first or `PANEL_EL` first, but don't swap them.

File: src/components/panel/panel.spec.js [hunk 12]
```diff
@@ -868,6 +895,87 @@ describe('$mdPanel', function() {
       });
     });
 
+    describe('CSS class logic:', function() {
+      it('should add a class to the container/wrapper when toElement=false',
+        function() {
```
Comment (ErinCoughlan) [inline src/components/panel/panel.spec.js]: nit: continuation indents are 4, not 2

File: src/components/panel/panel.spec.js [hunk 13]
```diff
@@ -1115,6 +1223,95 @@ describe('$mdPanel', function() {
...
+        myButton = '<button>myButton</button>';
+        attachToBody(myButton);
+        myButton = angular.element(document.querySelector('button'));
+        myButtonRect = myButton[0].getBoundingClientRect();
+      });
+
+      it('between two absolute positions', function() {
+        var top = '50px';
+        var left = '30px';
+        var position = $mdPanel.newPanelPosition().absolute().top(top).left(left);
```
Comment (ErinCoughlan) [inline src/components/panel/panel.spec.js]: nit: these will be more readable if you chain them on the next lines, like the relative positioning setup below.

File: src/components/panel/panel.spec.js [hunk 14]
```diff
@@ -1115,6 +1223,95 @@ describe('$mdPanel', function() {
...
+        config['position'] = position;
```
Comment (ErinCoughlan) [inline src/components/panel/panel.spec.js]: nit: new line before this one

**Comment by ErinCoughlan** [comment]
@ThomasBurleson @bradrich LGTM once nits are resolved

**Comment by bradrich** [comment]
@ErinCoughlan & @ThomasBurleson All requests have been met and everything should be GTG. Please let me know if I have missed anything.

Thanks for everyone's attention.
